middleware/kubernetes/handler.go

```diff
@@ -53,38 +51,46 @@ func (k Kubernetes) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.M
 			if (dns.CountLabel(name) - 1) < k.AutoPath.NDots {
 				break
 			}
-			// Search "svc.cluster.local" and "cluster.local"
+			origQName := state.QName()
+			// Search "svc.cluster.local." and "cluster.local."
 			for i := 0; i < 2; i++ {
 				path = strings.Join(dns.SplitDomainName(path)[1:], ".")
-				state = state.NewWithQuestion(strings.Join([]string{name, path}, "."), state.QType())
-				records, extra, _, err = k.routeRequest(zone, state)
+				newstate := state.NewWithQuestion(strings.Join([]string{name, path}, "."), state.QType())
+				records, extra, _, err = k.routeRequest(zone, newstate)
 				if !k.IsNameError(err) {
+					records = append(records, newCNAME(origQName, records[0].Header().Name, records[0].Header().Ttl))
 					break
 				}
 			}
 			if !k.IsNameError(err) {
 				break
 			}
-			// Fallthrough with the host search path (if set)
-			wr := rewrite.NewResponseReverter(w, r)
+			// Try host search path (if set) in the next middleware
+			apw := NewAutoPathWriter(w, r)
 			for _, hostsearch := range k.AutoPath.HostSearchPath {
-				r = state.NewWithQuestion(strings.Join([]string{name, hostsearch}, "."), state.QType()).Req
-				rcode, nextErr := middleware.NextOrFailure(k.Name(), k.Next, ctx, wr, r)
-				if rcode == dns.RcodeSuccess {
+				newstate := state.NewWithQuestion(strings.Join([]string{name, hostsearch}, "."), state.QType())
+				rcode, nextErr := middleware.NextOrFailure(k.Name(), k.Next, ctx, apw, newstate.Req)
+				if apw.Sent {
 					return rcode, nextErr
 				}
 			}
 			// Search . in this middleware
-			state = state.NewWithQuestion(strings.Join([]string{name, "."}, ""), state.QType())
-			records, extra, _, err = k.routeRequest(zone, state)
+			newstate := state.NewWithQuestion(strings.Join([]string{name, "."}, ""), state.QType())
+			records, extra, _, err = k.routeRequest(zone, newstate)
 			if !k.IsNameError(err) {
+				records = append(records, newCNAME(origQName, records[0].Header().Name, records[0].Header().Ttl))
```

Comment:
you need a put the CNAME first and then the original A record otherwise glibc won't understand it.